File: android/src/test/kotlin/com/hydrapayments/sdk/ModelTests.kt
package com.hydrapayments.sdk

import org.json.JSONObject
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

class ModelTests {
    // ── Account ──

    @Test
    fun accountFullDeserialization() {
        val json = """{"id":"acc_1","owner_id":"owner_123","account_type":"business","currency":"GBP","balance":"10000","metadata":{"ref":"ABC"},"created_at":"2024-01-01T00:00:00Z","updated_at":"2024-01-02T00:00:00Z"}"""
        val account = Account.fromJson(JSONObject(json))
        assertEquals("acc_1", account.id)
        assertEquals("owner_123", account.ownerId)
        assertEquals("business", account.accountType)
        assertEquals("GBP", account.currency)
        assertEquals("10000", account.balance)
        assertEquals("ABC", account.metadata?.get("ref"))
        assertEquals("2024-01-01T00:00:00Z", account.createdAt)
        assertEquals("2024-01-02T00:00:00Z", account.updatedAt)
    }

    @Test
    fun accountOptionalsDefaultToNull() {
        val json = """{"id":"acc_2","owner_id":"o2","account_type":"personal","currency":"USD","balance":"500","created_at":"2024-01-01T00:00:00Z"}"""
        val account = Account.fromJson(JSONObject(json))
        assertNull(account.metadata)
        assertNull(account.updatedAt)
    }

    // ── Transaction ──

    @Test
    fun transactionFullDeserialization() {
        val json = """{"id":"tx_1","source_account_id":"acc_src","dest_account_id":"acc_dst","amount":"1000","currency":"GBP","status":"completed","transaction_type":"transfer","reference":"ref_001","description":"payment","metadata":{"key":"val"},"previous_state_hash":"abc123","created_at":"2024-01-01T00:00:00Z","updated_at":"2024-01-02T00:00:00Z"}"""
        val tx = Transaction.fromJson(JSONObject(json))
        assertEquals("tx_1", tx.id)
        assertEquals("acc_src", tx.sourceAccountId)
        assertEquals("acc_dst", tx.destAccountId)
        assertEquals("1000", tx.amount)
        assertEquals("completed", tx.status)
        assertEquals("transfer", tx.transactionType)
        assertEquals("ref_001", tx.reference)
        assertEquals("payment", tx.description)
        assertEquals("val", tx.metadata?.get("key"))
        assertEquals("abc123", tx.previousStateHash)
    }

    @Test
    fun transactionOptionalsDefaultToNull() {
        val json = """{"id":"tx_2","amount":"500","currency":"USD","status":"pending","created_at":"2024-01-01T00:00:00Z"}"""
        val tx = Transaction.fromJson(JSONObject(json))
        assertNull(tx.sourceAccountId)
        assertNull(tx.destAccountId)
        assertNull(tx.reference)
    }

    // ── Wallet ──

    @Test
    fun walletCustodial() {
        val json = """{"id":"w_1","owner_id":"o1","wallet_type":"custodial","chain":"ethereum","address":"0xabc","is_custodial":true,"encrypted_private_key":"enc_key","created_at":"2024-01-01T00:00:00Z","updated_at":"2024-01-02T00:00:00Z"}"""
        val wallet = Wallet.fromJson(JSONObject(json))
        assertEquals("w_1", wallet.id)
        assertTrue(wallet.isCustodial)
        assertEquals("enc_key", wallet.encryptedPrivateKey)
    }

    @Test
    fun walletNonCustodial() {
        val json = """{"id":"w_2","owner_id":"o2","wallet_type":"non-custodial","chain":"solana","address":"0xdef","is_custodial":false,"created_at":"2024-01-01T00:00:00Z"}"""
        val wallet = Wallet.fromJson(JSONObject(json))
        assertEquals("w_2", wallet.id)
        assertFalse(wallet.isCustodial)
        assertNull(wallet.encryptedPrivateKey)
        assertNull(wallet.updatedAt)
    }

    // ── SplitRule ──

    @Test
    fun splitRuleWithNestedSplits() {
        val json = """{"id":"split_1","transaction_id":"tx_1","total":"1000","currency":"GBP","splits":[{"account_id":"acc_1","percentage":"70"},{"account_id":"acc_2","percentage":"30"}],"sink_account_id":"sink_1","status":"active","created_at":"2024-01-01T00:00:00Z"}"""
        val split = SplitRule.fromJson(JSONObject(json))
        assertEquals("split_1", split.id)
        assertEquals("tx_1", split.transactionId)
        assertEquals("1000", split.total)
        assertEquals(2, split.splits.size)
        assertEquals("acc_1", split.splits[0].accountId)
        assertEquals("70", split.splits[0].percentage)
        assertEquals("acc_2", split.splits[1].accountId)
        assertEquals("30", split.splits[1].percentage)
        assertEquals("sink_1", split.sinkAccountId)
        assertEquals("active", split.status)
    }

    // ── Gateway: Token ──

    @Test
    fun createTokenResponseWithNestedCard() {
        val json = """{"id":"tok_1","card":{"brand":"visa","last4":"1111","exp_month":"12","exp_year":"2025"},"created_at":"2024-01-01T00:00:00Z"}"""
        val token = CreateTokenResponse.fromJson(JSONObject(json))
        assertEquals("tok_1", token.id)
        assertEquals("visa", token.card?.brand)
        assertEquals("1111", token.card?.last4)
        assertEquals("12", token.card?.expMonth)
        assertEquals("2025", token.card?.expYear)
    }

    @Test
    fun createIntentResponse() {
        val json = """{"id":"pi_1","status":"requires_payment_method","amount":"2000","currency":"USD","client_secret":"sec_123"}"""
        val intent = CreateIntentResponse.fromJson(JSONObject(json))
        assertEquals("pi_1", intent.id)
        assertEquals("requires_payment_method", intent.status)
        assertEquals("2000", intent.amount)
        assertEquals("USD", intent.currency)
        assertEquals("sec_123", intent.clientSecret)
    }

    @Test
    fun createRefundResponse() {
        val json = """{"id":"ref_1","status":"succeeded","amount":"1000","charge":"ch_1"}"""
        val refund = CreateRefundResponse.fromJson(JSONObject(json))
        assertEquals("ref_1", refund.id)
        assertEquals("succeeded", refund.status)
        assertEquals("1000", refund.amount)
        assertEquals("ch_1", refund.charge)
    }

    // ── Health ──

    @Test
    fun healthResponseHealthy() {
        val json = """{"status":"healthy","version":"1.0","database":"connected"}"""
        val health = HealthResponse.fromJson(JSONObject(json))
        assertEquals("healthy", health.status)
        assertTrue(health.isHealthy)
        assertEquals("1.0", health.version)
        assertEquals("connected", health.database)
    }

    @Test
    fun healthResponseUnhealthy() {
        val json = """{"status":"unhealthy","database":"disconnected"}"""
        val health = HealthResponse.fromJson(JSONObject(json))
        assertEquals("unhealthy", health.status)
        assertFalse(health.isHealthy)
    }

    // ── Commission ──

    @Test
    fun commissionResponse() {
        val json = """{"total_commission":"1500"}"""
        val commission = CommissionResponse.fromJson(JSONObject(json))
        assertEquals("1500", commission.totalCommission)
    }

    // ── Webhook ──

    @Test
    fun webhookResponse() {
        val json = """{"received":true}"""
        val webhook = WebhookResponse.fromJson(JSONObject(json))
        assertTrue(webhook.received)
    }
}

File: ios/Tests/HydraPaymentsTests/ModelTests.swift
import XCTest
@testable import HydraPayments

final class ModelTests: XCTestCase {
    let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        return d
    }()

    // MARK: Account

    func testAccountFullDeserialization() throws {
        let json = """
        {"id":"acc_1","owner_id":"owner_123","account_type":"business","currency":"GBP","balance":"10000","metadata":{"ref":"ABC"},"created_at":"2024-01-01T00:00:00Z","updated_at":"2024-01-02T00:00:00Z"}
        """
        let account = try decoder.decode(Account.self, from: Data(json.utf8))
        XCTAssertEqual(account.id, "acc_1")
        XCTAssertEqual(account.ownerId, "owner_123")
        XCTAssertEqual(account.accountType, "business")
        XCTAssertEqual(account.currency, "GBP")
        XCTAssertEqual(account.balance, "10000")
        XCTAssertEqual(account.metadata?["ref"], "ABC")
        XCTAssertEqual(account.createdAt, "2024-01-01T00:00:00Z")
        XCTAssertEqual(account.updatedAt, "2024-01-02T00:00:00Z")
    }

    func testAccountOptionalsDefaultToNil() throws {
        let json = """
        {"id":"acc_2","owner_id":"o2","account_type":"personal","currency":"USD","balance":"500","created_at":"2024-01-01T00:00:00Z"}
        """
        let account = try decoder.decode(Account.self, from: Data(json.utf8))
        XCTAssertNil(account.metadata)
        XCTAssertNil(account.updatedAt)
    }

    // MARK: Transaction

    func testTransactionFullDeserialization() throws {
        let json = """
        {"id":"tx_1","source_account_id":"acc_src","dest_account_id":"acc_dst","amount":"1000","currency":"GBP","status":"completed","transaction_type":"transfer","reference":"ref_001","description":"payment","metadata":{"key":"val"},"previous_state_hash":"abc123","created_at":"2024-01-01T00:00:00Z","updated_at":"2024-01-02T00:00:00Z"}
        """
        let tx = try decoder.decode(Transaction.self, from: Data(json.utf8))
        XCTAssertEqual(tx.id, "tx_1")
        XCTAssertEqual(tx.sourceAccountId, "acc_src")
        XCTAssertEqual(tx.destAccountId, "acc_dst")
        XCTAssertEqual(tx.amount, "1000")
        XCTAssertEqual(tx.currency, "GBP")
        XCTAssertEqual(tx.status, "completed")
        XCTAssertEqual(tx.transactionType, "transfer")
        XCTAssertEqual(tx.reference, "ref_001")
        XCTAssertEqual(tx.description, "payment")
        XCTAssertEqual(tx.metadata?["key"], "val")
        XCTAssertEqual(tx.previousStateHash, "abc123")
    }

    func testTransactionOptionalsDefaultToNil() throws {
        let json = """
        {"id":"tx_2","amount":"500","currency":"USD","status":"pending","created_at":"2024-01-01T00:00:00Z"}
        """
        let tx = try decoder.decode(Transaction.self, from: Data(json.utf8))
        XCTAssertNil(tx.sourceAccountId)
        XCTAssertNil(tx.destAccountId)
        XCTAssertNil(tx.reference)
    }

    // MARK: Wallet

    func testWalletCustodial() throws {
        let json = """
        {"id":"w_1","owner_id":"o1","wallet_type":"custodial","chain":"ethereum","address":"0xabc","is_custodial":true,"encrypted_private_key":"enc_key","created_at":"2024-01-01T00:00:00Z","updated_at":"2024-01-02T00:00:00Z"}
        """
        let wallet = try decoder.decode(Wallet.self, from: Data(json.utf8))
        XCTAssertEqual(wallet.id, "w_1")
        XCTAssertTrue(wallet.isCustodial)
        XCTAssertEqual(wallet.encryptedPrivateKey, "enc_key")
    }

    func testWalletNonCustodial() throws {
        let json = """
        {"id":"w_2","owner_id":"o2","wallet_type":"non-custodial","chain":"solana","address":"0xdef","is_custodial":false,"created_at":"2024-01-01T00:00:00Z"}
        """
        let wallet = try decoder.decode(Wallet.self, from: Data(json.utf8))
        XCTAssertEqual(wallet.id, "w_2")
        XCTAssertFalse(wallet.isCustodial)
        XCTAssertNil(wallet.encryptedPrivateKey)
        XCTAssertNil(wallet.updatedAt)
    }

    // MARK: SplitRule

    func testSplitRuleWithNestedSplits() throws {
        let json = """
        {"id":"split_1","transaction_id":"tx_1","total":"1000","currency":"GBP","splits":[{"account_id":"acc_1","percentage":"70"},{"account_id":"acc_2","percentage":"30"}],"sink_account_id":"sink_1","status":"active","created_at":"2024-01-01T00:00:00Z"}
        """
        let split = try decoder.decode(SplitRule.self, from: Data(json.utf8))
        XCTAssertEqual(split.id, "split_1")
        XCTAssertEqual(split.transactionId, "tx_1")
        XCTAssertEqual(split.total, "1000")
        XCTAssertEqual(split.splits.count, 2)
        XCTAssertEqual(split.splits[0].accountId, "acc_1")
        XCTAssertEqual(split.splits[0].percentage, "70")
        XCTAssertEqual(split.splits[1].accountId, "acc_2")
        XCTAssertEqual(split.splits[1].percentage, "30")
        XCTAssertEqual(split.sinkAccountId, "sink_1")
        XCTAssertEqual(split.status, "active")
    }

    // MARK: Gateway

    func testCreateTokenResponseWithNestedCard() throws {
        let json = """
        {"id":"tok_1","card":{"brand":"visa","last4":"1111","exp_month":"12","exp_year":"2025"},"created_at":"2024-01-01T00:00:00Z"}
        """
        let token = try decoder.decode(CreateTokenResponse.self, from: Data(json.utf8))
        XCTAssertEqual(token.id, "tok_1")
        XCTAssertEqual(token.card?.brand, "visa")
        XCTAssertEqual(token.card?.last4, "1111")
        XCTAssertEqual(token.card?.expMonth, "12")
        XCTAssertEqual(token.card?.expYear, "2025")
    }

    func testCreateIntentResponse() throws {
        let json = """
        {"id":"pi_1","status":"requires_payment_method","amount":"2000","currency":"USD","client_secret":"sec_123"}
        """
        let intent = try decoder.decode(CreateIntentResponse.self, from: Data(json.utf8))
        XCTAssertEqual(intent.id, "pi_1")
        XCTAssertEqual(intent.status, "requires_payment_method")
        XCTAssertEqual(intent.amount, "2000")
        XCTAssertEqual(intent.currency, "USD")
        XCTAssertEqual(intent.clientSecret, "sec_123")
    }

    func testCreateRefundResponse() throws {
        let json = """
        {"id":"ref_1","status":"succeeded","amount":"1000","charge":"ch_1"}
        """
        let refund = try decoder.decode(CreateRefundResponse.self, from: Data(json.utf8))
        XCTAssertEqual(refund.id, "ref_1")
        XCTAssertEqual(refund.status, "succeeded")
        XCTAssertEqual(refund.amount, "1000")
        XCTAssertEqual(refund.charge, "ch_1")
    }

    // MARK: Health

    func testHealthResponseHealthy() throws {
        let json = """
        {"status":"healthy","version":"1.0","database":"connected"}
        """
        let health = try decoder.decode(HealthResponse.self, from: Data(json.utf8))
        XCTAssertEqual(health.status, "healthy")
        XCTAssertTrue(health.isHealthy)
        XCTAssertEqual(health.version, "1.0")
        XCTAssertEqual(health.database, "connected")
    }

    func testHealthResponseUnhealthy() throws {
        let json = """
        {"status":"unhealthy","database":"disconnected"}
        """
        let health = try decoder.decode(HealthResponse.self, from: Data(json.utf8))
        XCTAssertEqual(health.status, "unhealthy")
        XCTAssertFalse(health.isHealthy)
    }

    // MARK: Commission

    func testCommissionResponse() throws {
        let json = """
        {"total_commission":"1500"}
        """
        let commission = try decoder.decode(CommissionResponse.self, from: Data(json.utf8))
        XCTAssertEqual(commission.totalCommission, "1500")
    }

    // MARK: Webhook

    func testWebhookResponse() throws {
        let json = """
        {"received":true}
        """
        let webhook = try decoder.decode(WebhookResponse.self, from: Data(json.utf8))
        XCTAssertTrue(webhook.received)
    }

    // MARK: ApiResponse

    func testApiResponseSuccess() throws {
        let json = """
        {"success":true,"data":{"id":"acc_1","owner_id":"o1","account_type":"b","currency":"GBP","balance":"0","created_at":"2024-01-01T00:00:00Z"},"error":null}
        """
        let response = try decoder.decode(ApiResponse<Account>.self, from: Data(json.utf8))
        XCTAssertTrue(response.success)
        XCTAssertNotNil(response.data)
        XCTAssertNil(response.error)
    }

    func testApiResponseError() throws {
        let json = """
        {"success":false,"data":null,"error":"Insufficient funds"}
        """
        let response = try decoder.decode(ApiResponse<Account>.self, from: Data(json.utf8))
        XCTAssertFalse(response.success)
        XCTAssertNil(response.data)
        XCTAssertEqual(response.error, "Insufficient funds")
    }
}
